🚀 3단계 - 리팩터링(매장 식사 주문)

hoon25 opened the pull request:
안녕하세요. 마지막 3단계까지 리뷰 감사합니다 🙇🙇🙇

## BoundedContext, AggregateRoot간 Entity, VO 참조를 어떻게 가져갈 것인가?
### BoundedContext, AggregateRoot 분리 현황
- `Menu` BoundedContext
  -  `Menu`, `MenuGroup` Aggregate Root를 가짐
- `Product` BoundedContext
  - `Product` Aggregate Root를 가짐
- `EatInOrder` BoundedContext
  - `EatInOrder`, `OrderTable` Aggregate Root를 가짐

### 1. 다른 BoundedContext 다른 AggregateRoot
도메인 계층에서 재정의한 VO를 사용한다
ex) `Menu`에서 `Product` 사용 시 `Product`를 재구성한 `MenuProductInfo` 사용
ex) `EatInOrder`에서 `Menu` 사용 시 `Menu`를 재구성한 `EatInOrderMenuInfo` 사용

### 2. 같은 BoundedContext 다른 AggregateRoot
도메인 계층에서 Entity를 공유한다.
ex) `EatInOrder`에서는 `OrderTable` 도메인을 공유해서 사용


### 생각정리
**`같은 BoundedContext` `다른 AggregateRoot`인 경우**에도 처음에는 VO로 재정의해서 사용했는데, 의존성 분리를 위해서 불필요한 코드가 많아진다는 생각이 들더라구요. 
**같은 BoundedContext끼리는 `Entity`와 `Repository`를 공유해서 사용**하고, **BoundedContext 분리 시점**에 VO와 별개의 ***Client를 사용하는 방식으로 타협하면 어떨까 생각했어요.

hoon25 committed:
docs(README.md): EatInOrder 모델링 정리

hoon25 committed:
feat(OrderTable): domain 계층 구현

hoon25 committed:
feat(OrderTable): application 계층 구현

hoon25 committed:
feat(EatInOrder): domain 계층 구현

hoon25 committed:
feat(EatInOrder): EatInOrder 생성 기능 추가

hoon25 committed:
feat(EatInOrder): EatInOrder accept, serve 기능 추가

hoon25 committed:
feat(EatInOrder): EatInOrder complete 기능 추가

hoon25 committed:
feat(OrderTable): empty로 변경시 eatInOrder확인 로직 추가

hoon25 committed:
refactor(EatInOrder): OrderTableInfo VO 삭제 후 entity 사용

Gomding commented on src/main/kotlin/kitchenpos/order/tobe/eatinorder/domain/OrderTable.kt:
```diff
@@ -0,0 +1,35 @@
+package kitchenpos.order.tobe.eatinorder.domain
```
domain 내부 패키지에 depth 를 추가해서 OrderTable 관련 코드를 모아놔도 좋을 것 같아요!

hoon25 replied:
반영했어요!

Gomding commented on README.md:
```diff
@@ -119,18 +119,19 @@ docker compose -p kitchenpos up -d
 
```
주문쪽 모델링도 작성하신게 있다면 추가해주셔도 좋을 것 같아요!

Gomding commented on src/main/kotlin/kitchenpos/order/tobe/eatinorder/domain/EatInOrder.kt:
```diff
@@ -0,0 +1,69 @@
+    @Column(name = "order_table_id")
+    val orderTableId: UUID
+
+) : Order() {
```
Order 를 상속받으신 부분이 궁금해요!
창훈님은 `매장식사주문 / 배달주문 / 포장주문`  바운디드 컨텍스트를 어떻게 구성하실 예정이신가요?

hoon25 replied:
`매장식사주문 / 배달주문 / 포장주문` 에 상관없이 `Order`목록을 전체조회하는 기능 때문에 상속으로 구현했어요.

- 대안
  - CQRS도입으로 조회 모델을 분리
  - 애플리케이션 레벨에서 데이터 조합

대안으로는 이런 방법들이 있을것 같고, 대안으로 진행한다면 상속없이 바운디드컨텍스트를 구성하는게 좋아보여요.

Gomding replied:
바운디드 컨텍스트 구성에 대해 질문드린 이유는
세 주문 유형을 각각의 바운디드 컨텍스트로 나눈다면 `Order` 로 인해 변경에 대한 전파가 일어난다고 생각했어요 :)
바운디드 컨텍스트는 독립적이어야하고, 통합 조회를 위해서 결합이 일어나서는 안된다는 관점에서요.

창훈님이 제안해주신 방법도 좋다고 생각해요 👍 👍 👍 

정리하면 중복이 있어도 바운디드 컨텍스트의 관점에 따라 차이가 있으니 (강의중에 피자 가게 테이블에 있는 피자와 쓰레기통에 있는 피자가 다르다는 예시)
도메인 모델은 각자 독립적으로 유지하는것이 좋다고 생각합니다 :)

Gomding commented on src/main/kotlin/kitchenpos/order/tobe/eatinorder/domain/EatInOrder.kt:
```diff
@@ -0,0 +1,69 @@
+            "완료할 수 있는 상태가 아닙니다."
+        }
+        status = EatInOrderStatus.COMPLETED
+        registerEvent(EatInOrderCompleteEvent(id))
```
같은 `BoundedContext` 다른 `AggregateRoot` 관계에서도 도메인 이벤트로 처리하신 이유가 궁금해요!
다른 방법으로는 도메인 서비스를 활용하는 방법도 있었을 것 같아서요 :)
창호님의 구현 기준이 궁금했어요!

hoon25 replied:
**`DomainService`사용 vs `DomainEvent` 사용** 은 생각하지 못했던 부분이네요 🙇 

**기본적**으로는 `DomainService`를 사용하고, **협력하는 AggregateRoot가 많아서 복잡**해지면, 도메인이벤트를 사용하는 방식으로 변경할 것 같아요.

hoon25 replied:
도메인서비스를 사용하는 방식으로 변경했는데, 그와 별개로 `OrderTableEmptyService`  **(OrderTable의 empty관련 도메인서비스)** 를 추가하니 고민이 생기네요.

저는 `OrderTable`와 결합도가 높아지지만, complete로직의 응집도를 높이기위해서 실행을 도메인에 맡겼는데, 민영님은 어떻게 코드를 작성하실지 궁금합니다.

Gomding commented on src/main/kotlin/kitchenpos/order/tobe/eatinorder/domain/EatInOrder.kt:
```diff
@@ -0,0 +1,69 @@
+    )
+    var status: EatInOrderStatus,
+
+    @Column(name = "order_table_id")
+    val orderTableId: UUID
```
애그리거트간에 간접 참조로 구현해주셨군요 👍 👍

Gomding reviewed (changes requested):
안녕하세요 창호님!
3단계도 꼼꼼하게 구현해주셨네요 👍 👍
몇 가지 코멘트 남겼으니 확인부탁드려요 😃 
궁금한 점 있으면 언제든 DM이나 코멘트 남겨주세요!

hoon25 committed:
refactor(EatInOrder): complete로직 리팩토링

- complete 시 도메인이벤트가 아닌 도메인서비스로 EmptyTable처리
- OrderTable에 대한 OrderTableEmptyService 추가

hoon25 committed:
docs(EatInOrder): EatInOrder에 대한 모델링 추가

hoon25 committed:
move(OrderTable): EatInOrder domain패키지 가독성을 위해 OrderTable 관련 패키지 분리

hoon25 reviewed (commented):
리뷰 반영했습니다! 
`EatInOrder`의 `complete`에서는 의도적으로 도메인이벤트가 아닌 도메인서비스를 사용하려고 해보았어요.

Gomding commented on src/main/kotlin/kitchenpos/order/tobe/eatinorder/application/EatInOrderService.kt:
```diff
@@ -0,0 +1,68 @@
+        val orderTable = orderTableRepository.findById(eatInOrder.orderTableId).orElseThrow {
+            NoSuchElementException("주문 테이블이 존재하지 않습니다.")
+        }
+        eatInOrder.complete(orderTable, orderTableEmptyService)
```
> 도메인서비스를 사용하는 방식으로 변경했는데, 그와 별개로 OrderTableEmptyService (OrderTable의 empty관련 도메인서비스) 를 추가하니 고민이 생기네요.
> 
> 저는 OrderTable와 결합도가 높아지지만, complete로직의 응집도를 높이기위해서 실행을 도메인에 맡겼는데, 민영님은 어떻게 코드를 작성하실지 궁금합니다.

도메인 스스로 상태를 변경하고
OrderTableEmptyService 에게 변경 가능 판단의 역할을 맡기는것이 좋았다고 생각해요!

저도 창호님과 동일한 방식으로 작성할 것 같아요 :)
OrderTableEmptyService 내부에서 OrderTable 을 조회해와서 상태변경을 진행하는것도 고민해봤는데
객체 스스로에게 처리를 지시하고 `OrderTableEmptyService` 는 정책에 대한 판단을 하는게 좋은 설계라고 생각했어요!!

hoon25 replied:
억지로 도메인에 로직을 집중하려했나? 라는 고민이 있었는데, 민영님도 같은 관점이라고 하니 안심이 되네요 👍

Gomding commented on src/main/kotlin/kitchenpos/order/tobe/eatinorder/domain/ordertable/OrderTable.kt:
```diff
@@ -0,0 +1,44 @@
+        orderTableOccupancy = orderTableOccupancy.occupied()
+    }
+
+    fun empty(orderTableEmptyService: OrderTableEmptyService) {
+        if (!orderTableEmptyService.canEmpty(this)) {
+            throw IllegalStateException("주문테이블에 완료되지않은 주문이 존재합니다.")
+        }
+        this.orderTableOccupancy = OrderTableOccupancy.EMPTY
+    }
+
+    fun emptyIfPossible(orderTableEmptyService: OrderTableEmptyService) {
+        if (orderTableEmptyService.canEmpty(this)) {
+            this.orderTableOccupancy = OrderTableOccupancy.EMPTY
+        }
+    }
```
👍 👍 👍

Gomding reviewed (changes requested):
안녕하세요 창훈님 🙇 
의견 공유해주셔서 감사해요 :)
다음 요청에는 머지할 것 같은데 추가로 궁금한 점 있으면 코멘트 남겨주세요!!
없다면 바로 재요청주셔도 됩니다 😃

hoon25 reviewed (commented):
민영님 덕분에 3단계까지 잘 마무리 할 수 있었네요 🙇🙇

끝까지 친절하게 리뷰해주셔서 감사했습니다!!

Gomding reviewed (approved):
안녕하세요 창훈님!
말씀드린대로 마지막 단계 이만 머지하겠습니다 :)
미션 완주 축하드립니다 🎉 🎉 
정말 고생많으셨어요  😄
